### PR TITLE
[BALDE-47] Implement our own assertions

### PR DESCRIPTION
**Pull request by BIGbadEL**

### CODE REVIEW

File: engine/src/utils/bald_assert.h [hunk 1]
```diff
@@ -0,0 +1,109 @@
+//
```
Comment (Blinkuu) [inline engine/src/utils/bald_assert.h]: I think bald_assert stuff should go into debug folder.

**Review by Blinkuu — CHANGES_REQUESTED** [review]

File: engine/src/core/app/layer_stack.cpp [hunk 2]
```diff
@@ -8,7 +8,7 @@ namespace Bald {
 
     LayerStack::LayerStack() {
         [[maybe_unused]] bool state = Init();
-        assert(state);
+        BALD_ASSERT(state, "LayerStack", "Failed to initialized LayerStack", state);
```
Comment (Blinkuu) [inline engine/src/core/app/layer_stack.cpp]: initialize*

File: engine/src/core/app/layer_stack.h [hunk 3]
```diff
@@ -225,29 +225,29 @@ namespace Bald {
 
     template<typename L, typename ... Args>
     void LayerStack::PushLayer(Args&& ... args) {
-        static_assert(std::is_base_of<Layer, L>::value, "Layer is not the base of L");
+        BALD_STATIC_ASSERT(static_cast<bool >(std::is_base_of<Layer, L>::value), "Layer is not the base of L");
```
Comment (Blinkuu) [inline engine/src/core/app/layer_stack.h]: Is casting really necessary here?

File: engine/src/core/app/layer_stack.h [hunk 4]
```diff
@@ -225,29 +225,29 @@ namespace Bald {
...
         m_ForAddition.emplace(m_ForAddition.begin() + m_ForAdditionAmount, new L{args ...});
         ++m_ForAdditionAmount;
         EventManager::Emit<LayerPushedEvent>();
     }
 
     template<typename L, typename ... Args>
     void LayerStack::PushOverlay(Args&& ... args) {
-        static_assert(std::is_base_of<Layer, L>::value, "Layer is not the base of L");
+        BALD_STATIC_ASSERT(static_cast<bool >(std::is_base_of<Layer, L>::value), "Layer is not the base of L");
```
Comment (Blinkuu) [inline engine/src/core/app/layer_stack.h]: Is casting really necessary here?

File: engine/src/core/app/layer_stack.h [hunk 5]
```diff
@@ -225,29 +225,29 @@ namespace Bald {
...
         m_ForAddition.emplace_back(new L{args ...});
         EventManager::Emit<LayerPushedEvent>();
     }
 
     template<typename L>
     void LayerStack::PopLayer() {
-        static_assert(std::is_base_of<Layer, L>::value, "Layer is not the base of L");
+        BALD_STATIC_ASSERT(static_cast<bool >(std::is_base_of<Layer, L>::value), "Layer is not the base of L");
```
Comment (Blinkuu) [inline engine/src/core/app/layer_stack.h]: Is casting really necessary here?

File: engine/src/core/debug/bald_assert.h [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+        /**
+         * @fn SourceInfo
+         * @brief Constructs new SourceInfo object.
+         * @param file [std::string] name of a file where assertion occurred.
+         * @param line [int] number of line where assertion occurred.
+         */
+
+        SourceInfo(std::string&& file, int line) :
+                _file(std::move(file)), _line(line) { };
+        const std::string _file; /**< name of a file where assertion occurred. >*/
```
Comment (Blinkuu) [inline engine/src/core/debug/bald_assert.h]: Fix naming to match our convention

File: engine/src/core/debug/bald_assert.h [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+
+#define BALD_ASSERT(...)
+#define BALD_ASSERT_IMPL(...) BALD_ASSERT_IMPL_VARS
+#define BALD_ASSERT_IMPL_VARS
+
+#endif
+
+
+
+
```
Comment (Blinkuu) [inline engine/src/core/debug/bald_assert.h]: Fix formatting.

**Review by Blinkuu — CHANGES_REQUESTED** [review]

File: engine/src/core/app/layer_stack.h (hunk 4 above)
Comment (BIGbadEL) [reply engine/src/core/app/layer_stack.h]: yes.

File: engine/src/core/app/layer_stack.h (hunk 5 above)
Comment (BIGbadEL) [reply engine/src/core/app/layer_stack.h]: yes

File: engine/src/core/app/layer_stack.h (hunk 4 above)
Comment (Blinkuu) [reply engine/src/core/app/layer_stack.h]: why?

File: engine/src/core/events/event_manager.h [hunk 8]
```diff
@@ -192,15 +193,15 @@ namespace Bald {
...
+        BALD_STATIC_ASSERT(static_cast<bool >(std::is_base_of<Event, T>::value), "Event is not the base of T");
 
         T* event = new T{args...};
         m_EventQueue.push_back(event);
     }
 
     template<class T>
     bool EventManager::IsEventInQueue() noexcept {
-        static_assert(std::is_base_of<Event, T>::value, "Event is not the base of T");
+        BALD_STATIC_ASSERT(static_cast<bool >(std::is_base_of<Event, T>::value), "Event is not the base of T");
```
Comment (Blinkuu) [inline engine/src/core/events/event_manager.h]: Remove all static casting and this will be ready to merge.

**Review by Blinkuu — CHANGES_REQUESTED** [review]

**Review by Blinkuu — APPROVED** [review]

**Pull request merged**